marlowe-test/src/Spec/Marlowe/Plutus/MList.hs

```diff
@@ -0,0 +1,305 @@
+--   "unionWith f Nil l = l" |
+--   "unionWith f l Nil = l"
+-- @
+unionWith :: Ord a => (b -> b -> b) -> MList a b -> MList a b -> MList a b
+unionWith f ((x, y) : t) ((x2, y2) : t2) =
+  if x < x2
+    then (x, y) : unionWith f t ((x2, y2) : t2)
+    else if x > x2
+      then (x2, y2) : unionWith f ((x, y) : t) t2
+      else (x, f y y2) : unionWith f t t2
+unionWith _f [] l = l
+unionWith _f l [] = l
+
+
+-- | Find with default for Isabelle `MList`.
+--
+-- @
+-- fun findWithDefault :: "'b \<Rightarrow> 'a \<Rightarrow> (('a::linorder) \<times> 'b) list \<Rightarrow> 'b" where
+--   "findWithDefault d k l = (case lookup k l of
+--                               None \<Rightarrow> d
+--                             | Some x \<Rightarrow> x)"
+-- @
+findWithDefault :: Ord a => b -> a -> MList a b -> b
+findWithDefault d k l =
+  case lookup k l of
+    Nothing -> d
+    Just x -> x
+
+
+-- | Run tests.
+tests :: TestTree
+tests =
+  testGroup "MList vs AssocMap"
+    [
+      testCase     "`empty` is equivalent"           checkEmpty
+    , testProperty "`null` is equivalent"            checkNull
+    , testProperty "`singleton` is equivalent"       checkSingleton
+    , testProperty "`insert` is equivalent"          checkInsert
+    , testProperty "`delete` is equivalent"          checkDelete
+    , testProperty "`lookup` is equivalent"          checkLookup
+    , testProperty "`member` is equivalent"          checkMember
+    , testProperty "`unionWith` is equivalent"       checkUnionWith
+    , testProperty "`findWithDefault` is equivalent" checkFindWithDefault
+    ]
+
+
+-- | Generate a sorted `MList` with no duplicates.
+arbitraryMList :: Gen (MList Integer [()])
+arbitraryMList = nubBy ((==) `on` fst) . sortBy (compare `on` fst) <$> arbitrary
+
+
+-- | Compare an `MList` to an `AssocMap`.
+equivalent :: Ord a => Eq b => MList a b -> AM.Map a b -> Bool
+equivalent mlist assocmap = mlist == sortBy (compare `on` fst) (AM.toList assocmap)
+
+
+-- | Compare `empty` for `MList` and `AssocMap`.
+checkEmpty :: Assertion
+checkEmpty = assertBool "Empty MList and AssocMap" $ (empty :: MList [()] Integer) `equivalent` AM.empty
+
+
+-- | Compare `empty` for `MList` and `AssocMap`.
+checkNull :: Property
+checkNull = property $ do
+  let
+    gen = do
+      mlist <- arbitraryMList
+      assocmap <- AM.fromList <$> shuffle mlist
+      pure (mlist, assocmap)
+  forAll gen
+    $ \(mlist, assocmap) -> (== empty) mlist == AM.null assocmap
+
+
+-- | Compare `empty` for `MList` and `AssocMap`.
+checkSingleton :: Property
+checkSingleton = property $ do
+  let
+    gen = do
+      a <- arbitrary :: Gen Integer
+      b <- arbitrary :: Gen [()]
+      pure (a, b)
+  forAll gen
+    $ \(a, b) -> [(a, b)] `equivalent` AM.singleton a b
+
+
+-- | Compare `empty` for `MList` and `AssocMap`.
+checkInsert :: Property
+checkInsert = property $ do
+  let
+    gen = do
+      mlist <- arbitraryMList
+      assocmap <- AM.fromList <$> shuffle mlist
```

Comment:
I see that these tests start from MList and shuffle the AssocMap, shouldn't we do the opposite? Start by AssocMap, remove duplicates and sort for MList? The examples provided by the auditors are mostly preoccupied with duplicate entries or unsorted entries in MList, in these tests we are not covering that.

The Isabelle theorems `delete_valid` and `insert_valid` already prove that if the MList is valid before insert/delete it will be valid after

Reply:
Thanks. Fixed in 9e068b6a60103f2a40604a7eaea57c5f29b59180.